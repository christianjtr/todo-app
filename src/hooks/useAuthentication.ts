
import { useEffect } from 'react';
import { useHistory } from "react-router-dom";
import Cookies from 'js-cookie';
import * as LoginServices from '../services/loginService';
import * as TaskServices from '../services/taskService';
interface useAuthenticationInterface {
    checkAuthToken: () => boolean;
    logIn: (credentials: LoginServices.LoginCredentials) => Promise<void>;
    logOut: () => Promise<void>;
} 

const useAuthentication = (): useAuthenticationInterface => {
    
    const history = useHistory();

    useEffect(() => {
        if(checkAuthToken()) history.push("/home");
    }, [history]);
    
    const checkAuthToken = (): boolean => {
        return !!Cookies.get('authToken');
    }

    const logIn = async (credentials: LoginServices.LoginCredentials): Promise<void> => {
        const { authToken } = await LoginServices.logIn(credentials);
        
        Cookies.set('authToken', authToken, { 
            sameSite: 'strict',
            expires: 7,
            // secure: true, <-- HTTPS Protocols only
        });
        
        if(checkAuthToken()) history.push("/home");
    }

    const logOut = async (): Promise<void> => {
        await LoginServices.logOut();
        
        if(checkAuthToken()) Cookies.remove('authToken');
        TaskServices.deleteTasks();
        
        history.push("/login");
    }

    return {
        checkAuthToken,
        logIn,
        logOut
    }
}

export default useAuthentication;
