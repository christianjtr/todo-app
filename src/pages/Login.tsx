import React from "react";

const Login: React.FC = (): React.ReactElement => {
    return (
        <section className="hero is-primary is-fullheight">
            <div className="hero-body">
                <div className="container">
                    <div className="columns is-centered">
                        <div className="column is-5-tablet is-4-desktop is-3-widescreen">
                            <form action="" className="box">
                                <div className="field">
                                    <label htmlFor="" className="label">
                                        Email
                                    </label>
                                    <div className="control has-icons-left">
                                        <input
                                            type="email"
                                            placeholder="e.g. bobsmith@gmail.com"
                                            className="input"
                                            required
                                        />
                                        <span className="icon is-small is-left">
                                            <i className="fa fa-envelope"></i>
                                        </span>
                                    </div>
                                </div>
                                <div className="field">
                                    <label htmlFor="" className="label">
                                        Password
                                    </label>
                                    <div className="control has-icons-left">
                                        <input
                                            type="password"
                                            placeholder="*******"
                                            className="input"
                                            required
                                        />
                                        <span className="icon is-small is-left">
                                            <i className="fa fa-lock"></i>
                                        </span>
                                    </div>
                                </div>
                                <div className="field">
                                    <label htmlFor="" className="checkbox">
                                        <input type="checkbox" />
                                        Remember me
                                    </label>
                                </div>
                                <div className="field">
                                    <button className="button is-success">
                                        Login
                                    </button>
                                </div>
                            </form>
                        </div>
                    </div>
                </div>
            </div>
        </section>
    );
};

export default Login;
