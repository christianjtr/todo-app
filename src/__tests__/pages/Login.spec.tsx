import React from "react";
import "@testing-library/jest-dom";
import { render, fireEvent, cleanup } from "@testing-library/react";
import { createRenderer } from "react-test-renderer/shallow";
import useAuthentication from "../../hooks/useAuthentication";

import Login from "../../pages/Login";

/**
 * Example on how to test a page view...
 */

jest.mock("../../hooks/useAuthentication");

describe("Login page component", () => {
    beforeAll(() => {
        (useAuthentication as jest.Mock) = jest.fn().mockReturnValue({
            logIn: () => jest.fn(),
        });
    });

    afterEach(() => {
        cleanup();
    });

    afterAll(() => {
        jest.clearAllMocks();
    });

    it("should render component properly", () => {
        const renderer = createRenderer();

        const component = renderer.render(<Login />);

        expect(component).toBeDefined();
        expect(renderer.getRenderOutput()).toMatchSnapshot();
    });

    it("should perform a login by filling in the form and clicking on submit button", () => {
        useAuthentication().logIn = jest.fn();

        const { container } = render(<Login />);

        expect(useAuthentication).toHaveBeenCalled();

        const usernameInputElement = container.querySelector("#username");
        const passwordInputElement = container.querySelector("#password");
        const logInBtn = container.querySelector("#submitBtn");

        fireEvent.change(usernameInputElement!, {
            target: { value: "username" },
        });

        fireEvent.change(passwordInputElement!, {
            target: { value: "12345678" },
        });

        fireEvent.click(logInBtn!);

        expect(useAuthentication().logIn).toHaveBeenCalled();
    });
});
