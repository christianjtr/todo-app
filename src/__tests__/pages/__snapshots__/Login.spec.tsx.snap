// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page component should render component properly 1`] = `
<section
  className="hero is-fullheight"
>
  <div
    className="hero-body"
  >
    <div
      className="container"
    >
      <div
        className="columns is-centered"
      >
        <div
          className="column is-5-tablet is-4-desktop is-3-widescreen"
        >
          <form
            className="box"
            id="loginForm"
            name="loginForm"
            onSubmit={[Function]}
          >
            <div
              className="is-flex is-flex-direction-column m-4"
            >
              <img
                alt="Logo"
                className="is-align-self-center m-2"
              />
              <span
                className="is-size-5"
              >
                My TODO App
              </span>
            </div>
            <div
              className="field"
            >
              <div
                className="control has-icons-left"
              >
                <input
                  className="input"
                  id="username"
                  minLength={3}
                  name="username"
                  placeholder="User"
                  type="text"
                />
                <span
                  className="icon is-small is-left"
                >
                  <i
                    aria-hidden="true"
                    className="fa fa-user"
                  />
                </span>
              </div>
            </div>
            <div
              className="field"
            >
              <div
                className="control has-icons-left"
              >
                <input
                  className="input"
                  id="password"
                  maxLength={16}
                  minLength={8}
                  name="password"
                  placeholder="Password"
                  type="password"
                />
                <span
                  className="icon is-small is-left"
                >
                  <i
                    aria-hidden="true"
                    className="fa fa-lock"
                  />
                </span>
              </div>
            </div>
            <div
              className="field"
            >
              <button
                className="button is-transparent"
                form="loginForm"
                id="submitBtn"
                name="submitBtn"
                type="submit"
              >
                Log in
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</section>
`;
